File: crates/core/src/engine/pi_hooks_move/registry.rs
//! What the registry an earlier kendex wrote will and will not give up,
//! and how wide the answer reaches.
//!
//! Two obstacles live here and they are not the same size. The document
//! itself may be one kendex cannot edit at all, which is in the way of
//! every hook needing an edit in it. Or one hook's own entry may be one
//! kendex cannot pick out, which is in that hook's way and nobody else's
//! — a sibling whose entry is exactly where its record says has nothing
//! to do with it.

use super::Found;
use super::{Identity, Registered, legacy_registration, look, registered};
use crate::harness::pi;
use crate::lock::LockEntry;
use crate::model::Scope;

/// Why nothing in the legacy registry can be given up at all, when
/// nothing can: it is a link kendex did not make, it could not be read,
/// or it holds an entry kendex has to take out in a shape its editor
/// cannot rewrite. The obstacle is the document, so it blocks every hook
/// that needs an edit in it — which is what makes this the scope-wide
/// half. Absence is no obstacle, and neither is a document holding
/// nobody's entries but somebody else's.
pub(super) fn registry_block(
    root: &std::path::Path,
    scope: &Scope,
    ours: &[&LockEntry],
) -> Option<String> {
    let path = pi::legacy_hook_registry(root);
    let say = |why: String| {
        Some(format!(
            "{} {why}, so nothing under the name pi reserved was retired — a hook's registration and the script it names have to go together",
            path.display()
        ))
    };
    match look(&path) {
        Found::Absent => return None,
        Found::Linked(_) => return say("is a link kendex did not create".to_owned()),
        Found::Unreadable(_, error) => return say(format!("could not be read ({error})")),
        Found::Plain(_) => {}
    }
    let entries = match crate::scan::hooks::read_registrations(&path) {
        Ok(entries) => entries,
        Err(message) => return say(format!("could not be read ({message})")),
    };
    // Whether anything of kendex's is in there to take out at all. An
    // entry kendex cannot pick out of the document is one hook's problem,
    // not the document's, and it holds that hook on its own below.
    let holds_ours = ours.iter().any(|entry| {
        matches!(
            registered(&entries, &legacy_registration(entry, scope, root)),
            Registered::Ours
        )
    });
    if !holds_ours {
        return None;
    }
    match crate::fs::read_if_exists(&path) {
        Err(error) => say(format!("could not be read ({error})")),
        Ok(text) => match serde_json::from_str::<serde_json::Value>(&text.unwrap_or_default()) {
            Ok(_) => None,
            Err(error) => say(format!(
                "holds an entry kendex has to take out but could not be parsed ({error})"
            )),
        },
    }
}

/// Why this hook's own registration under the reserved name is not
/// kendex's to take, when it is not: the record points at one entry and
/// the document does not hold exactly that one — moved to another event
/// or matcher, or carried twice. That is evidence about this hook and no
/// other, so it holds this hook and no other; a sibling whose own entry
/// is exactly where its record says still moves.
///
/// Identity has to resolve before anything is removed. The edit takes out
/// every handler answering to it, so one kendex cannot tell from another
/// is held, not guessed at.
pub(super) fn registration_conflict(
    root: &std::path::Path,
    scope: &Scope,
    entry: &LockEntry,
) -> Option<String> {
    let path = pi::legacy_hook_registry(root);
    let entries = crate::scan::hooks::read_registrations(&path).ok()?;
    let legacy = legacy_registration(entry, scope, root);
    let command = &legacy.command;
    let say = |why: String| {
        Some(format!(
            "{} {why} — that entry is not kendex's to take, so this hook stays where it is; move it back, or take it out yourself",
            path.display()
        ))
    };
    match registered(&entries, &legacy) {
        Registered::Absent => None,
        // Found, and only once — but found is not gone. What the removal
        // will really leave behind is read back before a single byte of
        // this hook's is planned for the trash.
        Registered::Ours => survives_its_own_removal(&path, &legacy).then(|| {
            format!(
                "{} writes {command} in a shape kendex cannot take it out of — a handler standing directly under its event, rather than inside a matcher group — so this hook stays where it is; take that entry out yourself, and the script goes with it on the next refresh",
                path.display()
            )
        }),
        Registered::Elsewhere => say(format!(
            "no longer registers {command} where kendex recorded it"
        )),
        // Only the new path's reading answers these two, and only about
        // its own document: the reserved name's entry is proven reachable
        // the same way a line above, and a link where it lives is what
        // `registry_block` is for.
        Registered::Unreachable | Registered::Linked => None,
        Registered::Ambiguous => say(format!(
            "registers {command} more than once, so kendex cannot tell its own entry from the others"
        )),
    }
}

/// Whether the document really gives this entry up — proven by taking it
/// out and reading the document back, never by the edit reporting that it
/// ran. A handler written directly under its event is a shape the edit
/// reaches past: it succeeds, removes nothing, and the script would then
/// go to the trash while what runs it stayed, pointing at a path with
/// nothing at it.
///
/// Anything this cannot establish reads as surviving. A document that
/// will not take the edit is one kendex cannot express this removal in,
/// which is the same answer by a shorter road.
fn survives_its_own_removal(path: &std::path::Path, identity: &Identity) -> bool {
    let Ok(Some(text)) = crate::fs::read_if_exists(path) else {
        return true;
    };
    let edit = crate::configedit::ConfigEdit::RemoveHook {
        event: identity.event.clone(),
        command: identity.command.clone(),
    };
    let Ok(after) = edit.apply(&text) else {
        return true;
    };
    crate::scan::hooks::registrations_text(&after)
        .is_ok_and(|entries| !matches!(registered(&entries, identity), Registered::Absent))
}

File: pi-extensions/pi-hooks/extensions/registry.ts
import { existsSync, readFileSync } from "node:fs";
import { resolve } from "node:path";

import { piUserDir } from "./config.js";

/**
 * The registry key a `PreToolUse` hook is rendered under. Pi has no per-hook
 * runner, so kendex restates each hook event as the listener Pi fires
 * (`crates/core/src/harness/caps.rs::pi_listener`) and keys the rendered
 * registry by that name. tests/registry.test.ts holds this to that map: a
 * rename on either side is a registry written under one key and read under
 * another, which is every hook silently off.
 */
export const TOOL_CALL_LISTENER = "tool_call";

/** One hook the rendered registry asks the carrier to run. */
export interface RegisteredHook {
	/** The registered command, run through a shell exactly as kendex wrote it. */
	command: string;
	/**
	 * The rendered guard's script name when the command names one, `""` when it
	 * does not — a command-bodied hook is the person's own words and has no
	 * script of ours behind it. The per-guard settings are keyed by this name.
	 */
	name: string;
	/**
	 * How a refusal names this hook. A command-bodied hook is named by where it
	 * is registered, never by its own text: that text is the person's, it can
	 * hold a credential written inline, and a reason reaches the model. The
	 * ordinal counts registrations in the file, not the ones this call matched,
	 * so the person can find the entry the reason names.
	 */
	label: string;
	/**
	 * The script that command runs, for a hook kendex rendered: the file under
	 * the same root the registration was read from, spawned directly.
	 *
	 * The command kendex writes for a project-scope hook spells its path
	 * `$(git rev-parse --show-toplevel)/.pi/…`, and git's answer is not always
	 * kendex's: a vendored checkout inside a project is its own git root while
	 * kendex renders — and this reads — at the project above it, and a project
	 * with no git at all has no answer to give. Both would run the wrong file
	 * or none, and the substitution is a shell expansion this never has to
	 * perform. So a hook of ours is spawned at the path the registry it came
	 * from anchors, which is where kendex wrote it. A command that is not ours
	 * has no such path and is run as written.
	 */
	script?: string;
	/**
	 * A rendered hook whose script no scope holds. Still refused, kendex having
	 * registered it — but under a reason naming the render and its repair,
	 * rather than bash's exit-127 text from a spawn that judged nothing.
	 */
	missing?: true;
	/** Milliseconds the registration asks for, from its `timeout` in seconds. */
	budgetMs?: number;
}

/** What the registries say about one tool call. */
export interface RegistryRead {
	hooks: RegisteredHook[];
	/**
	 * A registry that exists and could not be read, named with its cause. The
	 * caller refuses on it: kendex labels these hooks enforced, and a file only
	 * kendex writes failing to parse is not the person standing their guards
	 * down.
	 */
	unreadable?: string;
	/**
	 * How many registrations the project's registry holds under this listener,
	 * where trust withheld them. Counted before the matcher, so the number is
	 * what the project has installed rather than what this one call happened
	 * to match — a notice given once has to be right the first time.
	 */
	withheld: number;
}

/**
 * Whether a registration's matcher covers the tool being called. Absent, empty
 * and `*` cover every tool, as they do for the Claude Code registry this shape
 * comes from; anything else is a whole-string regex, compared against the tool
 * said in Claude's own words so the pattern is read exactly as its author
 * wrote it (`vocab.ts`).
 *
 * A pattern that will not compile judges the call rather than skipping it. It
 * is a matcher kendex registered and labels enforced, and the alternative is a
 * guard that is silently off for every tool. Nothing here says which matcher
 * failed to compile: the hook runs, so its own verdict is what the person
 * sees. The flags are Claude Code's — none — so a matcher that compiles and
 * matches there compiles and matches here.
 */
function matches(matcher: unknown, toolName: string): boolean {
	if (typeof matcher !== "string") return true;
	const pattern = matcher.trim();
	if (pattern === "" || pattern === "*") return true;
	try {
		return new RegExp(`^(?:${pattern})$`).test(toolName);
	} catch {
		return true;
	}
}

/**
 * The guard name a registered command runs, or `""` where the command is not
 * one kendex wrote for a hook of its own.
 *
 * `engine::targets::pi_hook` registers a rendered script as the whole command
 * `bash "<path>/kendex/hooks/<name>.sh"` at both scopes, so the whole command
 * is what is read: a custom hook that merely mentions such a path — grepping
 * for one, say — is the person's own command and must run as written rather
 * than spawn the script it names. tests/registry.test.ts fills the renderer's
 * own templates and holds this to them; a render this stops recognising falls
 * to the shell lane, which is where a command of the person's already goes.
 */
export function renderedName(command: string): string {
	return command.trim().match(/^bash "(?:[^"]*\/)?kendex\/hooks\/([^/"]+)\.sh"$/)?.[1] ?? "";
}

/** A `readFileSync` failure that means the file is simply not there. */
function absent(error: unknown): boolean {
	const code = (error as { code?: unknown } | null)?.code;
	return code === "ENOENT" || code === "ENOTDIR";
}

/**
 * The registrations one scope root holds for a listener, in file order.
 * Position is counted before the matcher and the shape check, so both numbers
 * describe the file rather than this one call: the label points at an entry a
 * person can go and read, and `withheld` says what the project installed.
 */
function readRegistry(root: string, listener: string, toolName: string): RegistryRead {
	const path = resolve(root, "hooks.json");
	let parsed: unknown;
	try {
		parsed = JSON.parse(readFileSync(path, "utf8"));
	} catch (error) {
		// No registry is kendex having installed no hook here, and that is the
		// only reading that allows the call. Anything else — a permission the
		// session does not have, a directory in the file's place, a document
		// that will not parse — is a registry that exists and did not answer,
		// and a guard that did not run does not stand aside.
		if (absent(error)) return { hooks: [], withheld: 0 };
		return { hooks: [], withheld: 0, unreadable: `${path}: ${(error as Error).message}` };
	}
	const groups = (parsed as { hooks?: Record<string, unknown> } | null)?.hooks?.[listener];
	if (!Array.isArray(groups)) return { hooks: [], withheld: 0 };

	const hooks: RegisteredHook[] = [];
	let position = 0;
	for (const group of groups) {
		const entry = group as { matcher?: unknown; hooks?: unknown };
		if (!Array.isArray(entry.hooks)) continue;
		const covers = matches(entry.matcher, toolName);
		for (const registration of entry.hooks) {
			position += 1;
			const hook = registration as { type?: unknown; command?: unknown; timeout?: unknown };
			if (!covers || hook.type !== "command" || typeof hook.command !== "string" || hook.command === "") continue;
			const timeout = typeof hook.timeout === "number" && Number.isFinite(hook.timeout) && hook.timeout > 0
				? hook.timeout * 1000
				: undefined;
			const name = renderedName(hook.command);
			hooks.push({
				command: hook.command,
				name,
				label: name === "" ? `custom hook ${position} in ${path}` : name,
				script: name === "" ? undefined : resolve(root, "hooks", `${name}.sh`),
				budgetMs: timeout,
			});
		}
	}
	return { hooks, withheld: position };
}

/**
 * Every hook the rendered registries ask for on this listener and this tool:
 * the project's `<project>/.pi/kendex/hooks.json` first, then the global
 * `<root-anchored PI_CODING_AGENT_DIR or ~/.pi/agent>/kendex/hooks.json`.
 * `project` is the caller's already-resolved project root, or `undefined`
 * where the session is in none.
 *
 * The registry is the render itself, not a model of it: what a hook runs is
 * the command kendex wrote there, which is how the same declaration reaches
 * Claude Code and Codex. That is the whole reason a custom hook — whose
 * command exists nowhere else — can run here at all.
 *
 * The project registry names commands the project ships, so `trusted` is Pi's
 * answer for this workspace: a clone the person has not trusted must not get
 * its own code run on the first tool call of the session. Untrusted, the
 * project scope contributes nothing but its count, and the global registry
 * still answers: the person's own hooks are not the project's. That count has
 * one consumer, so `countWithheld` false skips the read entirely — work with
 * no consumer, on a file whose size an untrusted party chooses.
 *
 * One installation of a guard runs once. Where both scopes register the same
 * rendered script the project's answers — unless its script is not on disk,
 * which is a broken registration, and the healthy copy at the next scope
 * answers rather than being shadowed by it. Two command-bodied hooks are two
 * hooks and both run, nothing but the command identifying them.
 */
export function registeredHooks(listener: string, toolName: string, project: string | undefined, trusted: boolean, countWithheld = true): RegistryRead {
	const hooks: RegisteredHook[] = [];
	const byName = new Map<string, number>();
	let unreadable: string | undefined;
	let withheld = 0;

	const answering: RegistryRead[] = [];
	if (project !== undefined && (trusted || countWithheld)) {
		const read = readRegistry(resolve(project, ".pi", "kendex"), listener, toolName);
		// Untrusted, this registry says only how many hooks were withheld.
		// Its failure to parse is not carried: refusing on it would let a
		// clone nobody has trusted stop every tool call in the session.
		if (trusted) answering.push(read);
		else withheld += read.withheld;
	}
	answering.push(readRegistry(resolve(piUserDir(), "kendex"), listener, toolName));

	for (const read of answering) {
		unreadable ??= read.unreadable;
		for (const hook of read.hooks) {
			if (hook.name === "") {
				hooks.push(hook);
				continue;
			}
			const at = byName.get(hook.name);
			if (at === undefined) {
				byName.set(hook.name, hooks.length);
				hooks.push(hook);
				continue;
			}
			// A registration whose render is gone judges nothing, so it must
			// not stand in front of a healthy copy at the next scope down.
			if (hooks[at]!.script !== undefined && !existsSync(hooks[at]!.script!) && hook.script !== undefined && existsSync(hook.script)) {
				hooks[at] = hook;
			}
		}
	}
	// Stat-ed once, at the end, rather than per scope above it.
	for (const [, at] of byName) {
		const hook = hooks[at]!;
		if (hook.script !== undefined && !existsSync(hook.script)) hooks[at] = { ...hook, missing: true };
	}
	return { hooks, withheld, ...(unreadable === undefined ? {} : { unreadable }) };
}
